src/sources/pipe.rs

```diff
@@ -0,0 +1,247 @@
+use std::fs::File;
+use std::io::BufRead;
+use std::os::raw::c_int;
+use std::os::unix::io::FromRawFd;
+use std::{io, thread};
+
+use async_stream::stream;
+use bytes::Bytes;
+use chrono::Utc;
+use codecs::{
+    decoding::{DeserializerConfig, FramingConfig},
+    StreamDecodingError,
+};
+use futures::{channel::mpsc, executor, SinkExt, StreamExt};
+use indoc::indoc;
+use tokio_util::{codec::FramedRead, io::StreamReader};
+use vector_config::configurable_component;
+use vector_core::ByteSizeOf;
+
+use crate::{
+    codecs::DecodingConfig,
+    config::{
+        log_schema, GenerateConfig, Output, Resource, SourceConfig, SourceContext,
+        SourceDescription,
+    },
+    internal_events::{BytesReceived, OldEventsReceived, StreamClosedError},
+    serde::default_decoding,
+    shutdown::ShutdownSignal,
+    SourceSender,
+};
+/// Configuration for the `pipe` source.
+#[configurable_component(source)]
+#[derive(Clone, Debug)]
+#[serde(deny_unknown_fields)]
+pub struct PipeConfig {
+    /// The maximum buffer size, in bytes, of incoming messages.
+    ///
+    /// Messages larger than this are truncated.
+    #[serde(default = "crate::serde::default_max_length")]
+    pub max_length: usize,
+
+    /// Overrides the name of the log field used to add the current hostname to each event.
+    ///
+    /// The value will be the current hostname for wherever Vector is running.
+    ///
+    /// By default, the [global `host_key` option](https://vector.dev/docs/reference/configuration//global-options#log_schema.host_key) is used.
+    pub host_key: Option<String>,
+
+    #[configurable(derived)]
+    pub framing: Option<FramingConfig>,
+
+    #[configurable(derived)]
+    #[serde(default = "default_decoding")]
+    pub decoding: DeserializerConfig,
+
+    /// The file descriptor number to read from.
+    pub fd: c_int,
+}
+
+inventory::submit! {
+    SourceDescription::new::<PipeConfig>("pipe")
+}
+
+impl GenerateConfig for PipeConfig {
+    fn generate_config() -> toml::Value {
+        toml::from_str(indoc! {r#"
+            fd = 500
+        "#})
+        .unwrap()
+    }
+}
+
+#[async_trait::async_trait]
+#[typetag::serde(name = "pipe")]
+impl SourceConfig for PipeConfig {
+    async fn build(&self, cx: SourceContext) -> crate::Result<super::Source> {
+        pipe_source(self.clone(), cx.shutdown, cx.out)
+    }
+
+    fn outputs(&self) -> Vec<Output> {
+        vec![Output::default(self.decoding.output_type())]
+    }
+
+    fn source_type(&self) -> &'static str {
+        "pipe"
+    }
+
+    fn resources(&self) -> Vec<Resource> {
+        vec![Resource::Fd(0)]
+    }
+
+    fn can_acknowledge(&self) -> bool {
+        false
+    }
+}
+
+pub fn pipe_source(
```

Comment:
The large majority of the body of this function is nearly identical to stdin source's `stdin_source()`

(https://github.com/vectordotdev/vector/blob/e888a09777f654a737ee045dc2ecaa268282f589/src/sources/stdin.rs#L96)

I think it would be achievable to extract the common code into a separate function and have both stdin_source() and pipe_source() call it.

Something like:

```Rust

pub fn file_descriptor_source<R>(mut buffer: R, host_key: Option<String>, shutdown: Shutdown, mut out: SourceSender, name: &str) -> crate::Result<super::Source> where R: Send + io::BufRead + 'static {
// ... the common code
}

pub fn pipe_source(config: PipeConfig,
    shutdown: ShutdownSignal,
    mut out: SourceSender,
    ) -> crate::Result<super::Source> {
    
    // let mut pipe = ... ;
    
    file_descriptor_source(pipe, self.host_key, shutdown, out, "fd");
}
```